### DISCUSSION TITLE
ESP32→Azure IoT Hub→データ取得の一連の流れ

### DISCUSSION THREAD

**Discussion by uecken** [1ft-seabass/ms-azure-hackathon-league-2021#16] [Q&A]
下記参考にAzure IoT Hubに接続するマイコン(ESP32)へコンパイルを行いたいが、
https://docs.microsoft.com/en-us/samples/azure-samples/esp32-iot-devkit-get-started/sample/

下記画面のどちらを選択すればいいのか分からない。
![image](https://user-images.githubusercontent.com/5786149/144706098-1c5e7f4e-4a7e-485e-9726-822eda4ff631.png)

どなたかご教授頂けると幸いです。

**Comment by 1ft-seabass** [comment]
明日朝に向けてちょっと調べてみますね！もしかすると、VSCode使うこのやり方ではなく、ひとつひとつ丁寧に設定していく方法も込みで模索するかもです。

**Reply by uecken** [reply]
ありがとうございます！
サンプルスクリプトにプライマリ接続文字列を埋め込み、PlatformIOにてビルドしESP32へ書き込むことでIoTHUB側で受信できました！
やったことを書いておきます！

[作業内容]
Tech Briefing: Azure IoT の基本 (前編) - Azure IoT Hub と Azure IoT Central 概説　を見ながら、
https://youtu.be/RhGqrJcWBiQ?t=621

プライマリ接続文字列を取得し、
![image](https://user-images.githubusercontent.com/5786149/144711532-7f1aa6c6-55f6-4209-b78a-a3f1189d7b2d.png)

https://docs.microsoft.com/en-us/samples/azure-samples/esp32-iot-devkit-get-started/sample/
のExampleを使い、
![image](https://user-images.githubusercontent.com/5786149/144711651-b7d40075-7d24-4d6f-9122-a8eac40598d0.png)


static const char* connectionString = "xxxxxxxxxxxxxxxxxx";
にプライマリ接続文字列を埋め込み、WiFi設定も行い、
Visual Studio PlatformIOでビルドしてESP32に書き込むこで、IoTHUB側で受信できました！
![image](https://user-images.githubusercontent.com/5786149/144711611-50ba2645-a27b-4062-a7b2-3413aff1f0aa.png)

ありがとうございます！

**Reply by 1ft-seabass** [reply]
おお！やはり！その方向ですよね！もしかすると、これで IoT Hub に送るという意味では解決しましたか？

**Reply by uecken** [reply]
はい！ありがとうございます！

今はIoTHUB→WebServerへデータを受け渡す方法を検討中です。
この辺でストレージに格納すれば、WebServer側から取ってこれるでしょうか？
![image](https://user-images.githubusercontent.com/5786149/144711946-dc28e71f-c0ac-4b0d-984f-4fe67328c91e.png)

**Reply by uecken** [reply]
今は「IoT Hub メッセージ ルーティングを使用して device-to-cloud メッセージを別のエンドポイントに送信する」を読んでいます。
https://docs.microsoft.com/ja-jp/azure/iot-hub/iot-hub-devguide-messages-d2c

**Reply by uecken** [reply]
「クイックスタート: IoT プラグ アンド プレイ デバイスから Azure IoT Hub にテレメトリを送信する」にて、
https://docs.microsoft.com/ja-jp/azure/iot-develop/quickstart-send-telemetry-iot-hub?pivots=programming-language-python


”デバイス サンプルを実行する”より、pythonでIoTHUBから情報取得できるようになりました。
```
kenji@DESKTOP-GVRLFUH:/mnt/c/Users/the_n/azure-iot-sdk-python/azure-iot-device/samples/pnp$ python3 temp_controller_with_thermostats.py
/usr/lib/python3/dist-packages/requests/__init__.py:89: RequestsDependencyWarning: urllib3 (1.26.7) or chardet (3.0.4) doesn't match a supported version!
  warnings.warn("urllib3 ({}) or chardet ({}) doesn't match a supported "
Connecting using Connection String HostName=lushboard.azure-devices.net;DeviceId=lushboard;SharedAccessKey=Ul0Kimp7nGJdajkCSDFJpsLaAfm9phlW2/4gD2rlNGI=
Updating pnp properties for root interface
{'serialNumber': 'alohomora'}
Updating pnp properties for thermostat1
{'thermostat1': {'maxTempSinceLastReboot': 98.34, '__t': 'c'}}
Updating pnp properties for thermostat2
{'thermostat2': {'maxTempSinceLastReboot': 48.92, '__t': 'c'}}
Updating pnp properties for deviceInformation
{'deviceInformation': {'swVersion': '5.5', 'manufacturer': 'Contoso Device Corporation', 'model': 'Contoso 4762B-turbo', 'osName': 'Mac Os', 'processorArchitecture': 'x86-64', 'processorManufacturer': 'Intel', 'totalStorage': 1024, 'totalMemory': 32, '__t': 'c'}}
Listening for command requests and property updates
Press Q to quit
Sending telemetry from various components
Sent message
{"temperature": 26}
```

**Reply by uecken** [reply]
IoTHUBに送信できただけでした。
ESP32→Azure IoT Hub→Backend　のBackendへデータ送信できていません。

**Reply by uecken** [reply]
下記参考になりそう。
「演習 - テレメトリを送信および受信するコードを記述する」
https://docs.microsoft.com/ja-jp/learn/modules/remotely-monitor-devices-with-azure-iot-hub/4-exercise-write-code-device-telemetry?pivots=vscode-node

**Reply by 1ft-seabass** [reply]
追いつきました！ちょっと見てみますね。

**Reply by 1ft-seabass** [reply]
@uecken 確認しました。おっしゃるとおり「テレメトリを受信するための 2 つ目のアプリを作成する」のやり方で合ってます！

Node.js か C# のコードがありますが、この言語で書けそうですかー？

**Reply by uecken** [reply]
やるならNode.jsですが、まだ手が出せていません..。
IoT HUBからAzure Cosmos DBにJSON形式で格納できたりしないでしょうか・・・？

**Reply by 1ft-seabass** [reply]
例えば、こういう連携どうでしょう？

IoT Hub > Functions > Azure Cosmos DB なやりかたとか。

https://ak1211.com/7704/

**Reply by uecken** [reply]
右下のイベントハブ接続にて、「利用できる既存の接続はありませ」と出てきてしまいます。
![image](https://user-images.githubusercontent.com/5786149/144734552-3b0584fc-2234-4cb2-a2a4-32eca016210b.png)

**Reply by 1ft-seabass** [reply]
なるほどなるほど。ちょっとお声がけします。

**Reply by uecken** [reply]
「M5Stack Core2で測定したデータをAzure Cosmos DBに蓄積してみた」をやってみましたが、
![image](https://user-images.githubusercontent.com/5786149/144735909-6a8af752-e1e1-4f93-a2a9-1506783ee6c2.png)

IoTHUBからデータを取得できていないようです。
kenji@DESKTOP-GVRLFUH:/mnt/c/Users/the_n/OneDrive/ドキュメント/Pln/OneDrive/ドキュメント/PlatformIO/ProjatformIO/Projects/RUSHBOARD/cheesecaveservice$ node app.js 
Cheese Cave Operator: the back-end service app
/mnt/c/Users/the_n/node_modules/azure-iot-common/lib/connection_string.js:29     
                throw new errors_1.ArgumentError(err + key);
                ^

ArgumentError: The connection string is missing the property: SharedAccessKeyName

**Reply by uecken** [reply]
#41 を参照。
